compiler/rustc/build.rs

```diff
@@ -0,0 +1,31 @@
+use std::path::PathBuf;
+use std::{env, fs};
+
+fn main() {
+    let target_os = env::var("CARGO_CFG_TARGET_OS");
+    let target_env = env::var("CARGO_CFG_TARGET_ENV");
+    if Ok("windows") == target_os.as_deref() && Ok("msvc") == target_env.as_deref() {
+        set_windows_exe_options();
+    }
+}
+
+// Add a manifest file to rustc.exe and increase the main thread's stack size.
+fn set_windows_exe_options() {
+    const STACK_SIZE: u32 = 2 * 1024 * 1024;
+    static WINDOWS_MANIFEST_FILE: &str = "Windows Manifest.xml";
+
+    let mut manifest = PathBuf::from(env::var_os("OUT_DIR").unwrap());
+    manifest.push(WINDOWS_MANIFEST_FILE);
+    fs::copy(WINDOWS_MANIFEST_FILE, &manifest).unwrap();
+
+    println!("cargo:rerun-if-changed={}", WINDOWS_MANIFEST_FILE);
+    // By default reserve `STACK_SIZE` many bytes for new thread, including the main thread).
+    // Note that memory is not actually committed until used.
+    println!("cargo:rustc-link-arg-bin=rustc-main=/STACK:{}", STACK_SIZE);
+    // Embed the Windows application manifest file.
+    println!("cargo:rustc-link-arg-bin=rustc-main=/MANIFEST:EMBED");
+    println!("cargo:rustc-link-arg-bin=rustc-main=/MANIFESTUAC");
```

Comment:
We don't need administrator permissions, so I don't think this is necessary.

Reply:
That option attempts to prevent Windows getting clever with auto elevation. The default option for `MANIFESTUAC` is `asInvoker` which [means](https://docs.microsoft.com/en-us/cpp/build/reference/manifestuac-embeds-uac-information-in-manifest?view=msvc-170):

> The application runs at the same permission level as the process that started it. You can elevate the application to a higher permission level by selecting Run as Administrator.

I'm uncertain how necessary it is in the case of rustc but I generally use it as it leads to more predictable behaviour.

Reply:
I've now made this more explicit and linked to the MSDN docs.

Reply:
If the UAC option isn't declared then Windows uses certain heuristics to decide if a binary should be elevated. This will still prompt the user with the UAC dialogue so it's not quite the same as the elevation that happens with some Microsoft signed executables.

But in any case, it's unwanted so we may as well make that explicit, no?

Reply:
According to https://docs.microsoft.com/en-us/windows/security/identity-protection/user-account-control/how-user-account-control-works the heuristics seem to be just for 32bit installer and uninstaller programs. I think all of the heuristics are deterministic in the sense that any given executable either triggers them or doesn't independent of the environment. Given that we haven't seen any issues about it being triggered I am pretty sure it will never trigger on rustc.exe.

Reply:
Fair enough, I can remove it. I guess the difference between "pretty sure" and "totally confident" isn't that big.